File: Sources/Scratchpad/Rendering/StrokeSmoothing.swift
/// One instruction in a smoothed stroke path. Kept as a plain, `Equatable`
/// value rather than building a `CGPath` directly so the smoothing math is
/// unit-testable without needing to introspect Core Graphics internals.
enum PathSegment: Equatable {
    case move(to: StrokePoint)
    case line(to: StrokePoint)
    case quadCurve(to: StrokePoint, control: StrokePoint)
}

/// Turns raw mouse-tracked points into a smoothed path by drawing a
/// quadratic curve through the midpoint of each consecutive pair, using the
/// shared point as the curve's control point. This is the standard
/// technique for smoothing freehand input without needing to buffer and
/// re-process the whole stroke on every new point.
enum StrokeSmoothing {
    static func segments(for points: [StrokePoint]) -> [PathSegment] {
        guard let first = points.first else { return [] }
        guard points.count > 1 else {
            // A single point (a tap, not a drag) still needs to render as a
            // dot: a zero-length line with a round cap draws one.
            return [.move(to: first), .line(to: first)]
        }

        var segments: [PathSegment] = [.move(to: first)]
        for index in 1..<(points.count - 1) {
            let current = points[index]
            let next = points[index + 1]
            let midpoint = StrokePoint(x: (current.x + next.x) / 2, y: (current.y + next.y) / 2)
            segments.append(.quadCurve(to: midpoint, control: current))
        }
        segments.append(.line(to: points[points.count - 1]))
        return segments
    }
}

File: Windows/Scratchpad/Rendering/StrokeSmoothing.cs
using Scratchpad.Drawing;

namespace Scratchpad.Rendering;

/// Turns raw mouse-tracked points into a smoothed path by drawing a
/// quadratic curve through the midpoint of each consecutive pair, using the
/// shared point as the curve's control point. This is the standard
/// technique for smoothing freehand input without needing to buffer and
/// re-process the whole stroke on every new point. Identical algorithm to
/// the macOS app's `StrokeSmoothing`.
public static class StrokeSmoothing
{
    public static IReadOnlyList<PathSegment> Segments(IReadOnlyList<StrokePoint> points)
    {
        if (points.Count == 0) return Array.Empty<PathSegment>();

        var first = points[0];
        if (points.Count == 1)
        {
            // A single point (a tap, not a drag) still needs to render as a
            // dot: a zero-length line with a round cap draws one.
            return new PathSegment[] { new PathSegment.Move(first), new PathSegment.Line(first) };
        }

        var segments = new List<PathSegment> { new PathSegment.Move(first) };
        for (var index = 1; index < points.Count - 1; index++)
        {
            var current = points[index];
            var next = points[index + 1];
            var midpoint = new StrokePoint((current.X + next.X) / 2, (current.Y + next.Y) / 2);
            segments.Add(new PathSegment.QuadCurve(midpoint, current));
        }
        segments.Add(new PathSegment.Line(points[^1]));
        return segments;
    }
}
